h11/_headers.py

```diff
@@ -153,12 +207,12 @@ def get_comma_header(headers, name):
 def set_comma_header(headers, name, new_values):
     # The header name `name` is expected to be lower-case bytes.
     new_headers = []
-    for found_name, found_raw_value in headers:
+    for found_raw_name, found_name, found_raw_value in headers._full_items:
         if found_name != name:
-            new_headers.append((found_name, found_raw_value))
+            new_headers.append((found_raw_name, found_raw_value))
     for new_value in new_values:
-        new_headers.append((name, new_value))
-    headers[:] = normalize_and_validate(new_headers)
+        new_headers.append((name.title(), new_value))
```

Comment:
I think this is my only concern, as it changes the h11 set header casing to title rather than lowercase. Which could be problematic if lowercase headers are desired (no way to avoid h11 writing titlecase now). Does you usecase require titlecased response headers as well? Alternatively, would simply knowing the raw casing be sufficient?

Reply:
So, `get_comma_header` and `set_comma_header` both expect lowercase headers.

We've got two options with this PR:

* Switch both `get_comma_header` and `set_comma_header` to expect case-sensitive headers, and lowercase them for the lookup behaviour. Preserve whatever header casing is used when calling `set_comma_header`.
* Continue to expect lowercase headers to both `get_comma_header` and `set_comma_header`. When setting up the header casing in `set_comma_header`, use title casing.

Of those two, I've gone for the second, as being marginally less impactful, and also being marginally quicker. (Only requires a casing transform in one of the two cases.)

Both of these two are private methods, so we're okay for them to have behaviour that only makes sense if they're being used in an appropriate context. In particular, the docstring for `get_comma_header` already lists the full set of header cases that it's actually being used with.

I've now added a corresponding docstring to `set_comma_header`, which I think helps clarify (internally), why the title casing is sufficient and appropriate for this function.

Reply:
I don't think I explained my concern well enough, however you have given answer I was seeking 

> we use title casing for the header names, in order to match the conventional HTTP header style.

For clarity my concern is that this choice means that I can't use h11 to exclusively write lowercase headers over the wire (equally with the current version I can't use h11 to exclusively write titlecased headers). 

I think this change is ok - I'll note it in the release notes and we can always add another flag if someone really needs to only write lowercase headers.

Reply:
I feel like I've seen cases where people complain about using lowercase'd headers because the server they were talking to didn't support them. Title cased is probably (sadly) marginally safer. I think preserving user-casing is probably the least surprising but also most expensive